core/remotes/docker/referrers.go

```diff
@@ -0,0 +1,122 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/containerd/errdefs"
+	"github.com/containerd/log"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+func (r dockerFetcher) FetchReferrers(ctx context.Context, dgst digest.Digest, artifactTypes ...string) ([]ocispec.Descriptor, error) {
+	rc, size, err := r.openReferrers(ctx, dgst, artifactTypes...)
+	if err != nil {
+		return nil, err
+	}
+	defer rc.Close()
+
+	var index ocispec.Index
+	dec := json.NewDecoder(io.LimitReader(rc, size))
+	if err := dec.Decode(&index); err != nil {
+		return nil, fmt.Errorf("failed to decode referrers index: %w", err)
+	}
+	if _, err := dec.Token(); !errors.Is(err, io.EOF) {
+		return nil, fmt.Errorf("unexpected data after JSON object")
+	}
+
+	if len(artifactTypes) == 0 {
+		return index.Manifests, nil
+	}
+
+	var referrers []ocispec.Descriptor
+	tFilter := map[string]struct{}{}
+	for _, t := range artifactTypes {
+		tFilter[t] = struct{}{}
+	}
+	for _, desc := range index.Manifests {
+		if _, ok := tFilter[desc.ArtifactType]; ok {
+			referrers = append(referrers, desc)
+		}
+	}
+	return referrers, nil
+}
+
+func (r dockerFetcher) openReferrers(ctx context.Context, dgst digest.Digest, artifactTypes ...string) (io.ReadCloser, int64, error) {
+	mediaType := ocispec.MediaTypeImageIndex
+	ctx = log.WithLogger(ctx, log.G(ctx).WithField("digest", dgst))
+
+	hosts := r.filterHosts(HostCapabilityReferrers)
+	var fallbackHosts []RegistryHost
+	if len(hosts) == 0 {
+		fallbackHosts = r.filterHosts(HostCapabilityResolve)
+		if len(fallbackHosts) == 0 {
+			return nil, 0, fmt.Errorf("no pull hosts: %w", errdefs.ErrNotFound)
```

Comment:
nit: seems like this is the error message from `fetcher.go` when the pull capability is filtered on. I'm not sure it has to be another word than "pull" but maybe should be differentiated for referrers?